### PR TITLE
Fix Preconditioner state corruption when nnx.Variable-wrapped params passed to init_fn

### PR DESCRIPTION
**Pull request by Merlin-Fang**
## Motivation
`flax.nnx` has become the recommended API for new Flax projects, and more
users are likely to reach for `nnx.Optimizer` going forward. This PR makes
`soap_jax.soap` work correctly when used that way, without changing any
behavior for existing Linen/`TrainState` users.

## Problem
When `scale_by_soap`'s `init_fn` receives params from `nnx.Optimizer` (as
`flax.nnx.Variable`-wrapped leaves rather than bare arrays), the resulting
`Preconditioner` state gets corrupted after the first optimizer step.

## Root cause
`nnx.Optimizer` passes params as `nnx.Param` (a `Variable` subclass). Without
unwrapping, `jtu.tree_map` traverses into `Param(array)` and reconstructs
`Param(Preconditioner)`. On the next step, `nnx.state` flattens `Preconditioner`
into `State({0: arr, 1: arr})`, destroying the original type.

## Fix
Strip `Variable` wrappers at the top of `init_fn` before building `GG`/`Q`.
Guarded by an `isinstance` check, so behavior with plain pytrees (e.g. Linen's
`TrainState`) is unchanged — `isinstance(p, Variable)` is `False` for bare
arrays, so params pass through untouched.

## Verification
Compared `soap` run via Linen `TrainState` vs. the same optimizer run via
`nnx.Optimizer`, same initial weights and inputs, 10 steps each. Predictions
match to `atol=1e-10`. 
```python
import jax
jax.config.update("jax_enable_x64", True)

import jax.numpy as jnp
from flax import linen as nn
from flax import nnx
from flax.training import train_state
from soap_jax.soap import soap

DTYPE = jnp.float64
KEY = jax.random.PRNGKey(0)
x = jnp.linspace(0, 2 * jnp.pi, 256, dtype=DTYPE)[:, None]
y = jnp.sin(x)

OPT_KWARGS = dict(
    learning_rate=3e-4, b1=0.95, b2=0.95,
    precondition_frequency=5, weight_decay=0.0, qr_dtype=DTYPE,
)


class MLPLinen(nn.Module):
    @nn.compact
    def __call__(self, x):
        x = nn.relu(nn.Dense(32, dtype=DTYPE, param_dtype=DTYPE)(x))
        x = nn.relu(nn.Dense(32, dtype=DTYPE, param_dtype=DTYPE)(x))
        return nn.Dense(1, dtype=DTYPE, param_dtype=DTYPE)(x)


class MLPNnx(nnx.Module):
    def __init__(self, rngs: nnx.Rngs, dtype=DTYPE):
        self.layer1 = nnx.Linear(1, 32, rngs=rngs, dtype=dtype)
        self.layer2 = nnx.Linear(32, 32, rngs=rngs, dtype=dtype)
        self.layer3 = nnx.Linear(32, 1, rngs=rngs, dtype=dtype)

    def __call__(self, x):
        x = nnx.relu(self.layer1(x))
        x = nnx.relu(self.layer2(x))
        return self.layer3(x)


# --- Linen reference, via train_state.TrainState ---
model_l = MLPLinen()
params = model_l.init(KEY, x)
state_l = train_state.TrainState.create(
    apply_fn=model_l.apply, params=params, tx=soap(**OPT_KWARGS),
)

# --- NNX model with identical initial weights, via nnx.Optimizer ---
model_n = MLPNnx(rngs=nnx.Rngs(0), dtype=DTYPE)
pp = params["params"]
model_n.layer1.kernel.value = pp["Dense_0"]["kernel"]
model_n.layer1.bias.value = pp["Dense_0"]["bias"]
model_n.layer2.kernel.value = pp["Dense_1"]["kernel"]
model_n.layer2.bias.value = pp["Dense_1"]["bias"]
model_n.layer3.kernel.value = pp["Dense_2"]["kernel"]
model_n.layer3.bias.value = pp["Dense_2"]["bias"]

optimizer_n = nnx.Optimizer(model_n, soap(**OPT_KWARGS), wrt=nnx.Param)

assert jnp.allclose(model_l.apply(params, x), model_n(x), atol=1e-10), \
    "initial params do not match between Linen and NNX"


# --- run both for 10 steps ---
def linen_loss_fn(p):
    return jnp.mean((model_l.apply(p, x) - y) ** 2)


def nnx_loss_fn(m):
    return jnp.mean((m(x) - y) ** 2)


for step in range(10):
    loss_l, grads_l = jax.value_and_grad(linen_loss_fn)(state_l.params)
    state_l = state_l.apply_gradients(grads=grads_l)

    loss_n, grads_n = nnx.value_and_grad(nnx_loss_fn)(model_n)
    optimizer_n.update(model_n, grads_n)  # <- raised AttributeError on step 2 before the fix

    print(f"step {step:>2}  linen_loss={float(loss_l):.6f}  nnx_loss={float(loss_n):.6f}")


# --- final numerical check ---
y_linen = model_l.apply(state_l.params, x)
y_nnx = model_n(x)
max_diff = float(jnp.max(jnp.abs(y_linen - y_nnx)))
print(f"\nmax diff after 10 steps: {max_diff:.2e}")

assert jnp.allclose(y_linen, y_nnx, atol=1e-10), f"FAIL  max_diff={max_diff:.2e}"
print("PASS  soap now matches across Linen TrainState and nnx.Optimizer  atol=1e-10")
```

### CODE REVIEW

**Comment by Merlin-Fang** [comment]
I came across #3, which addresses the same underlying issue (`nnx.Optimizer` breaking `Preconditioner` state) via a larger architectural change (replacing the wrapper with a plain pytree). This PR takes a more minimal approach: a 4-line unwrap at the `init_fn` boundary that leaves the existing `Preconditioner` class and Linen behavior untouched.
Happy to close this if #3 is the direction you'd rather take.

**Comment by haydn-jones** [comment]
I like this more than #3 (simply because its a smaller change), though I don't know if I like depending directly on flax. Is there a way we can handle this *without* importing flax? 

Also, I don't think flax is in our deps list. Not sure if it's a transitive dependency of what we have, but assuming we went with this approach we'd want to add it to the list explicitly.

**Comment by Merlin-Fang** [comment]
Good point! Just pushed a change that gates the flax import behind a try/except, so `soap_jax` doesn't hard-depend on flax:
```python
    try:
        from flax.nnx.variablelib import Variable as NnxVariable
    except ImportError:
        NnxVariable = None
```
Verified both directions:
- With flax installed: numerical match to Linen `TrainState` still holds (`atol=1e-10`, same test as before).
- Without flax installed (fresh env, no flax in site-packages): ran `scripts/test_soap_mlp.py` end-to-end with Equinox, loss drops as expected, no flax-related errors anywhere.

On the "avoid importing flax entirely" question: I don't think there's a way around it if we want proper support for `nnx.Optimizer`. NNX's `Variable` wrapping is structural, and the only reliable way to identify and unwrap it is `isinstance` against the actual class, and that requires importing it from flax. A pure duck-typing check like `hasattr(p, "get_value")`) might risk false positives from unrelated objects that coincide with the method name.

So I think gating the import behind try/except is the safest and cleanest way to get NNX support and zero additional dependency for users who don't need it.

**Comment by haydn-jones** [comment]
Yeah ok you're right. This is an issue with flax (imo). Can we do something like this to be a bit more version agnostic?

```python
try:
    from flax.nnx import Variable as NnxVariable
except ImportError:
    NnxVariable = None

def _unwrap(p):
    # flax >= 0.12 Variable has get_value(); 0.11 Variable only has .value
    return p.get_value() if hasattr(p, "get_value") else p.value
```

And also can you add `chex` as a dependency (its no longer transitive through optax). No need to add flax.

**Comment by Merlin-Fang** [comment]
Great idea! Pushed both changes.

One caveat: I tried to locally verify the `flax <= 0.11` branch but ran into an environment issue unrelated to this fix. flax 0.11 imports `mutable_array` from `jax._src.core`, which isn't present in any jax release currently on PyPI (latest is 0.10.2), so I couldn't get a working flax==0.11 install to test against. The `.value` branch mirrors the `get_value()` branch exactly in structure, so I'm reasonably confident it's correct, but wanted to flag it's untested rather than claim otherwise. Let me know if you have a known-good way to get flax 0.11 running, happy to verify further.

`get_value()` branch (flax 0.12.7) is verified as before, plus a clean run of `scripts/test_soap_mlp.py` in an environment with no flax installed at all.

File: src/soap_jax/soap.py [hunk 1]
```diff
@@ -11,6 +11,11 @@
 from jaxtyping import Array
 from optax import GradientTransformation, Updates
 
+try:
```
Comment (haydn-jones) [inline src/soap_jax/soap.py:14]: Can you use the public export `from flax.nnx import Variable`

**Review by haydn-jones — CHANGES_REQUESTED** [review]

**Comment by Merlin-Fang** [comment]
Good catch! Just changed it.

**Comment by haydn-jones** [comment]
Thanks!
